dask_expr/shuffle.py

```diff
@@ -691,17 +696,23 @@ def _meta(self):
 
 
 class _SetIndexPostScalar(Blockwise):
-    _parameters = ["frame", "index_name", "drop"]
+    _parameters = ["frame", "index_name", "drop", "new_divisions"]
```

Comment:
I'm uncomfortable putting divisions explicitly into operands.  It makes things somewhat brittle and hard to change I think.

Can this class not look up the expression tree to determine these values?

https://github.com/dask-contrib/dask-expr/issues/186

Reply:
Only ``SetPartition`` and ``SetIndex`` are aware of the divisions, since the calculation happens there. But both classes aren't part of the final expression tree after the simplify steps since they are removed in ``_simplify_down``. I think we have to pass this information along in ``_simplify_down`` in some way.

A related note: This will move into ``_lower`` as soon as it is available, so we would lock that in later on during the optimization process.

Reply:
Will put up a doc PR shortly.

``SetIndex`` will make the decision if we have to shuffle our DataFrame or if we are already sorted so that a simple blockwise ``set_index`` followed by a blockwise ``sort_index`` is sufficient. So we have to know the divisions to make that decision (for now). This is implemented in #206

We could (theoretically) move this to compute time, but this would significantly restrict follow up operations right now (e.g. set_index followed by a merge, merge wouldn't have a clue that a blockwise merge might be sufficient).

We could also go in a different direction and use a "``set_index`` is dumb" approach, but this would place a bigger burden on our users to give us the necessary information

I completely agree that this should come as late as possible, but I think we have to do this in ``optimize`` (for now at least). 

This describes the current situation. I think there are some options available that we can look into. For example, ``SetIndex`` could inspect the expression tree to check whether knowing the divisions right now makes sense or if we can defer to compute time. But that's not what I am aiming for right now. I'd like to get ``set_index`` running as good as possible without going down any rabbit holes.